### PR TITLE
[don't merge] add option to detour to skip loopback and LAN address

### PR DESCRIPTION
**Pull request by fffw**
It resolves https://github.com/getlantern/lantern/issues/2875 as well as https://github.com/getlantern/lantern/issues/2915 hopefully. Also refactored detour a bit and fixed the randomly failed test.

There's still one minor issue left: if the loopback or LAN address is unable to dial for some reason, `detour.Dialer` will not return immediately, so caller will wait until it times out. Hope I can fix it tomorrow.

### CODE REVIEW

File: src/github.com/getlantern/detour/detour.go [hunk 1]
```diff
@@ -137,11 +147,23 @@ func Dialer(detourDialer dialFunc) func(network, addr string) (net.Conn, error)
 
 		// dialing sequence
 		if whitelisted(addr) {
-			dialDetour(network, addr, detourDialer, ch)
+			go func() {
```
Comment (fffw) [inline src/github.com/getlantern/detour/detour.go]: moved all goroutines to `detour.go` to make the intention more explicit.

File: src/github.com/getlantern/detour/detour.go [hunk 2]
```diff
@@ -137,11 +147,23 @@ func Dialer(detourDialer dialFunc) func(network, addr string) (net.Conn, error)
...
+				dialDetour(network, addr, detourDialer, ch)
+			}()
 		} else {
+			tt := time.AfterFunc(TimeoutToConnect, func() {
```
Comment (fffw) [inline src/github.com/getlantern/detour/detour.go]: extract the timer from the loop below (L187 previously)

**Comment by myleshorton** [comment]
Ha very interesting on the ad blocker front @fffw -- I'll have a look at this thanks.

**Comment by myleshorton** [comment]
Oh looks like one of the detour tests is failing @fffw

**Comment by fffw** [comment]
This should also resolve https://github.com/getlantern/lantern/issues/3319.

Seems the test is not reliable enough...

**Comment by myleshorton** [comment]
Wait so is this one now obsolete @fffw?

**Comment by fffw** [comment]
Yeah indeed though lots of code can be reused. Mark it as don't merge until new detour merged.

**Comment by fffw** [comment]
This PR doesn't bypass LAN addresses if user chooses "Proxy all traffic". It will also expose user's access attempts via DNS queries.

One alternative would be to have user supply a list of sites to bypass proxy. It adds complexity on UI but provides a way to user for special requirements: local development sites, slow sites but must accessed inland (https://github.com/getlantern/lantern/issues/3325). It will not resolve https://github.com/getlantern/lantern/issues/2875 though.

**Pull request closed without merging**